elfi/methods/utils.py

```diff
@@ -184,29 +184,72 @@ def logpdf(cls, x, means, cov=1, weights=None):
         return np.log(cls.pdf(x, means=means, cov=cov, weights=weights))
 
     @classmethod
-    def rvs(cls, means, cov=1, weights=None, size=1, random_state=None):
+    def rvs(cls, means, cov=1, weights=None, size=1, prior_logpdf=None,
+            max_trials=100, random_state=None):
         """Draw random variates from the distribution.
 
         Parameters
         ----------
         means : array_like
             means of the Gaussian mixture components
-        weights : array_like
-            1d array of weights of the gaussian mixture components
-        cov : array_like
+        cov : array_like, optional
             a shared covariance matrix for the mixture components
-        size : int or tuple
+        weights : array_like, optional
+            1d array of weights of the gaussian mixture components
+        size : int or tuple, optional
+        prior_logpdf : callable, optional
+            Can be used to check validity of random variable.
+        max_trials : int, optional
+            Maximum number of trials to find allowed random variables.
         random_state : np.random.RandomState or None
 
         """
-        means, weights = cls._normalize_params(means, weights)
         random_state = random_state or np.random
+        means, weights = cls._normalize_params(means, weights)
 
-        inds = random_state.choice(len(means), size=size, p=weights)
-        rvs = means[inds]
-        perturb = ss.multivariate_normal.rvs(
-            mean=means[0] * 0, cov=cov, random_state=random_state, size=size)
-        return rvs + perturb
+        # additional complexity due to supporting scipy-like ambiguity of arguments
+        if size is None:
+            size = means.shape[0]
+            output = []
+            scalar = True
+        else:
+            output = np.empty((size,) + means.shape[1:])
+            scalar = False
+
+        n_accepted = 0
+        n_left = size
+        trials = 0
+
+        while n_accepted < size:
+            inds = random_state.choice(len(means), size=n_left, p=weights)
+            rvs = means[inds]
+            perturb = ss.multivariate_normal.rvs(mean=means[0]*0,
+                                                 cov=cov,
+                                                 random_state=random_state,
+                                                 size=n_left)
+            x = rvs + perturb
+
+            # check validity of x
+            if prior_logpdf is not None:
+                x = x[np.isfinite(prior_logpdf(x))]
+                n_accepted1 = len(x)
+            else:
+                n_accepted1 = size
+
+            output[n_accepted: n_accepted+n_accepted1] = x
+            n_accepted += n_accepted1
+            n_left -= n_accepted1
```

Comment:
I think this approach has the danger of being slow. If the probability of getting a particle from within the prior support is small, then it requires a substantial amount of iterations before the last few samples are finally acquired. Sophisticated method would track the acceptance ratio and estimate the amount of needed particle proposals. A simpler one could just repeat the original size until we have enough. The excess particles can be just discarded.

Reply:
Possibly, but it's also possible that sampling is expensive and should not be done more than needed. In any case my testing suggests this doesn't cause any performance penalty at least for our example models.

Reply:
Maybe the warning of very many iterations done would suffice for now.